### PR TITLE
feat(pricing): add price rule entity

### PR DESCRIPTION
**Pull request by pKorsholm**
**What**
- add price-rule entity to pricing module

blocked by #4977 

Fixes CORE-1497

### CODE REVIEW

File: packages/types/src/pricing/service.ts [hunk 1]
```diff
@@ -153,5 +157,41 @@ export interface IPricingModuleService {
     sharedContext?: Context
   ): Promise<RuleTypeDTO[]>
 
-  deleteRuleTypes(ruleTypes: string[], sharedContext?: Context): Promise<void>
+  deleteRuleTypes(
+    ruleTypes: string[],
```
Comment (riqwan) [inline packages/types/src/pricing/service.ts]:
```suggestion
    ruleTypeIds: string[],
```

File: packages/pricing/src/services/price-rule.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+  private buildQueryForList(
+    filters: PricingTypes.FilterablePriceRuleProps = {},
+    config: FindConfig<PricingTypes.PriceRuleDTO> = {}
+  ) {
+    const queryOptions = ModulesSdkUtils.buildQuery<PriceRule>(filters, config)
+
+    if (filters.id) {
+      queryOptions.where.id = { $in: filters.id }
+    }
```
Comment (riqwan) [inline packages/pricing/src/services/price-rule.ts]:
```suggestion
```

File: packages/pricing/src/services/pricing-module.ts [hunk 3]
```diff
@@ -61,6 +66,8 @@ export default class PricingModuleService<
     this.moneyAmountService_ = moneyAmountService
     this.ruleTypeService_ = ruleTypeService
     this.priceSetService_ = priceSetService
+    this.ruleTypeService_ = ruleTypeService
```
Comment (riqwan) [inline packages/pricing/src/services/pricing-module.ts]:
```suggestion
```

File: packages/pricing/src/services/pricing-module.ts [hunk 4]
```diff
@@ -520,4 +527,102 @@ export default class PricingModuleService<
...
     await this.ruleTypeService_.delete(ruleTypes, sharedContext)
   }
+
+  @InjectManager("baseRepository_")
+  async retrievePriceRule(
+    id: string,
+    config: FindConfig<PricingTypes.PriceRuleDTO> = {},
+    @MedusaContext() sharedContext: Context = {}
+  ): Promise<PricingTypes.PriceRuleDTO> {
+    const ruleType = await this.priceRuleService_.retrieve(
```
Comment (riqwan) [inline packages/pricing/src/services/pricing-module.ts]:
```suggestion
    const priceRule = await this.priceRuleService_.retrieve(
```

File: packages/pricing/src/services/pricing-module.ts [hunk 5]
```diff
@@ -520,4 +527,102 @@ export default class PricingModuleService<
...
+      id,
+      config,
+      sharedContext
+    )
+
+    return this.baseRepository_.serialize<PricingTypes.PriceRuleDTO>(ruleType, {
```
Comment (riqwan) [inline packages/pricing/src/services/pricing-module.ts]:
```suggestion
    return this.baseRepository_.serialize<PricingTypes.PriceRuleDTO>(priceRule, {
```

File: packages/pricing/src/services/pricing-module.ts [hunk 6]
```diff
@@ -520,4 +527,102 @@ export default class PricingModuleService<
...
+    })
+  }
+
+  @InjectManager("baseRepository_")
+  async listPriceRules(
+    filters: PricingTypes.FilterablePriceRuleProps = {},
+    config: FindConfig<PricingTypes.PriceRuleDTO> = {},
+    @MedusaContext() sharedContext: Context = {}
+  ): Promise<PricingTypes.PriceRuleDTO[]> {
+    const ruleTypes = await this.priceRuleService_.list(
```
Comment (riqwan) [inline packages/pricing/src/services/pricing-module.ts]:
```suggestion
    const priceRules = await this.priceRuleService_.list(
```

File: packages/pricing/src/services/pricing-module.ts [hunk 7]
```diff
@@ -520,4 +527,102 @@ export default class PricingModuleService<
...
+      filters,
+      config,
+      sharedContext
+    )
+
+    return this.baseRepository_.serialize<PricingTypes.PriceRuleDTO[]>(
+      ruleTypes,
```
Comment (riqwan) [inline packages/pricing/src/services/pricing-module.ts]:
```suggestion
      priceRules,
```

File: packages/pricing/src/services/pricing-module.ts [hunk 8]
```diff
@@ -520,4 +527,102 @@ export default class PricingModuleService<
...
+    return [
+      await this.baseRepository_.serialize<PricingTypes.PriceRuleDTO[]>(
+        ruleTypes,
+        {
+          populate: true,
+        }
+      ),
+      count,
+    ]
+  }
```
Comment (riqwan) [inline packages/pricing/src/services/pricing-module.ts]:
```suggestion
    const [priceRules, count] = await this.priceRuleService_.listAndCount(
      filters,
      config,
      sharedContext
    )

    return [
      await this.baseRepository_.serialize<PricingTypes.PriceRuleDTO[]>(
        priceRules,
        {
          populate: true,
        }
      ),
      count,
    ]
  }
```

File: packages/pricing/src/services/pricing-module.ts [hunk 9]
```diff
@@ -520,4 +527,102 @@ export default class PricingModuleService<
...
+    @MedusaContext() sharedContext: Context = {}
+  ): Promise<PricingTypes.PriceRuleDTO[]> {
+    const ruleTypes = await this.priceRuleService_.create(data, sharedContext)
+
+    return this.baseRepository_.serialize<PricingTypes.PriceRuleDTO[]>(
+      ruleTypes,
+      {
+        populate: true,
+      }
+    )
```
Comment (riqwan) [inline packages/pricing/src/services/pricing-module.ts]:
```suggestion
    const priceRules = await this.priceRuleService_.create(data, sharedContext)

    return this.baseRepository_.serialize<PricingTypes.PriceRuleDTO[]>(
      priceRules,
      {
        populate: true,
      }
    )
```

File: packages/pricing/src/services/pricing-module.ts [hunk 10]
```diff
@@ -520,4 +527,102 @@ export default class PricingModuleService<
...
+    @MedusaContext() sharedContext: Context = {}
+  ): Promise<PricingTypes.PriceRuleDTO[]> {
+    const ruleTypes = await this.priceRuleService_.update(data, sharedContext)
+
+    return this.baseRepository_.serialize<PricingTypes.PriceRuleDTO[]>(
+      ruleTypes,
+      {
+        populate: true,
+      }
+    )
```
Comment (riqwan) [inline packages/pricing/src/services/pricing-module.ts]:
```suggestion
    const priceRules = await this.priceRuleService_.update(data, sharedContext)

    return this.baseRepository_.serialize<PricingTypes.PriceRuleDTO[]>(
      priceRules,
      {
        populate: true,
      }
    )
```

**Review by riqwan — APPROVED** [review]
Added some suggestions, lgtm otherwise.

File: packages/pricing/src/models/price-rule.ts [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+import MoneyAmount from "./money-amount"
+import PriceSet from "./price-set"
+import PriceSetMoneyAmount from "./price-set-money-amount"
+import RuleType from "./rule-type"
+import { generateEntityId } from "@medusajs/utils"
+
+type OptionalFields = "id" | "is_dynamic" | "priority" 
+type OptionalRelations = "price_set" | "rule_type" | "price_set_money_amount"
+@Entity()
```
Comment (adrien2p) [inline packages/pricing/src/models/price-rule.ts:23]: nit: new line to get some breath

File: packages/pricing/src/models/price-rule.ts [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+export default class PriceRule {
+  [OptionalProps]: OptionalFields | OptionalRelations
+  @PrimaryKey({ columnType: "text" })
```
Comment (adrien2p) [inline packages/pricing/src/models/price-rule.ts:27]: nit: new line

File: packages/pricing/src/models/price-rule.ts [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+  @Property({ columnType: "integer", default: 0 })
+  priority: number
+
+  @ManyToOne({ entity: () => PriceSetMoneyAmount, fieldName: "price_set_money_amount_id", name: "price_set_money_amount_id_unique" })
+  price_set_money_amount: PriceSetMoneyAmount
+  
+  @Property({ columnType: "text" })
+  price_list_id!: string
+
+  // TODO: Add price list
```
Comment (adrien2p) [inline packages/pricing/src/models/price-rule.ts:63]: question: will it be part of a different pr?

File: packages/pricing/src/models/price-rule.ts [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+  
+  @BeforeCreate()
+  onCreate() {
```
Comment (adrien2p) [inline packages/pricing/src/models/price-rule.ts]: todo: beforeCreate to prevent conflict if we have an after create

File: packages/pricing/src/repositories/index.ts [hunk 15]
```diff
@@ -3,3 +3,4 @@ export { CurrencyRepository } from "./currency"
 export { MoneyAmountRepository } from "./money-amount"
 export { PriceSetRepository } from "./price-set"
 export { RuleTypeRepository } from "./rule-type"
+export { PriceRuleRepository } from "./price-rule"
```
Comment (adrien2p) [inline packages/pricing/src/repositories/index.ts]: nit: missing new line, I think globally a lint would be nice

File: packages/pricing/src/repositories/price-rule.ts [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+        ruleData.price_set_id
+      )
+
+      ruleDataClone.price_set_money_amount = await manager.findOne(
+        PriceSetMoneyAmount,
+        { id: ruleData.price_set_money_amount_id }
+      ) as PriceSetMoneyAmount
+
+      return ruleDataClone
+    }))
```
Comment (adrien2p) [inline packages/pricing/src/repositories/price-rule.ts]: todo(non blocking): I think it would be better to map first to create the clones, then findAll price set money amount and attach them on the values of the map in order to reduce to only one query instead of X

File: packages/pricing/src/services/price-rule.ts (hunk 2 above)
Comment (adrien2p) [reply packages/pricing/src/services/price-rule.ts]: therefore this method can be removed

**Review by adrien2p — APPROVED** [review]
LGTM, few comments here and there but nothing blocking

File: packages/pricing/src/repositories/price-rule.ts (hunk 16 above)
Comment (riqwan) [reply packages/pricing/src/repositories/price-rule.ts:97]: fair warning, this will be changed in the calculate pricing PR. None of this will be happening in that.

File: packages/pricing/src/models/price-rule.ts (hunk 13 above)
Comment (pKorsholm) [reply packages/pricing/src/models/price-rule.ts:63]: supposedly, atm not part of this pr, price list functionality is also not the focus of developing the pricing module at the moment. Keeping this until we eventually start messing with the interactions of price-sets and price lists would be the better approach to not overcomplicate something that's already pretty complicated 😅

File: packages/pricing/src/repositories/price-rule.ts (hunk 16 above)
Comment (pKorsholm) [reply packages/pricing/src/repositories/price-rule.ts]: there's a much neater trick with mikro-orm that Riqwan pointed out to me after this pr was opened. You don't need references or anything: 

```
      ruleDataClone.rule_type = ruleData.rule_type_id

      ruleDataClone.price_set = ruleData.price_set_id

      ruleDataClone.price_set_money_amount = ruleData.price_set_money_amount_id 
```
